Extension/src/LanguageServer/Providers/callHierarchyProvider.ts

```diff
@@ -124,76 +130,54 @@ export class CallHierarchyProvider implements vscode.CallHierarchyProvider {
-                        const requestStatus: CallHierarchyRequestStatus = canceledByUser ? CallHierarchyRequestStatus.CaneledByUser : CallHierarchyRequestStatus.Canceled;
-                        this.logTelemetry(CallHierarchyCallsToEvent, requestStatus, progressBarDuration);
-                        reject(new vscode.CancellationError());
-                    } else {
-                        this.logTelemetry(CallHierarchyCallsToEvent, CallHierarchyRequestStatus.Succeeded, progressBarDuration);
-                        if (result?.calls.length === 0) {
-                            resolve(undefined);
-                        } else {
-                            resolve(this.createIncomingCalls(result.calls));
-                        }
-                    }
-
-                    this.client.clearPendingReferencesCancellations();
-                };
-
-                workspaceReferences.setCallHierarchyResultsCallback(resultCallback);
-                workspaceReferences.startCallHierarchyIncomingCalls(params);
-
-                token.onCancellationRequested(e => {
-                    if (params === DefaultClient.referencesParams) {
-                        this.client.cancelReferences();
-                    }
-                });
-            };
-
-            if (DefaultClient.referencesRequestPending || workspaceReferences.symbolSearchInProgress) {
-                const cancelling: boolean = DefaultClient.referencesPendingCancellations.length > 0;
-                DefaultClient.referencesPendingCancellations.push({
-                    reject: () => {
-                        // Complete with nothing instead of rejecting, to avoid an error message from VS Code
-                        resolve(undefined);
-                    }, callback
-                });
-                if (!cancelling) {
-                    workspaceReferences.referencesCanceled = true;
-                    this.client.languageClient.sendNotification(CancelReferencesNotification);
-                }
-            } else {
-                callback();
-            }
-        });
+        Promise<vscode.CallHierarchyIncomingCall[] | undefined> {
+        await this.client.awaitUntilLanguageClientReady();
+        workspaceReferences.cancelCurrentReferenceRequest(CancellationSender.NewRequest);
+
+        const CallHierarchyCallsToEvent: string = "CallHierarchyCallsTo";
+        if (item === undefined) {
+            this.logTelemetry(CallHierarchyCallsToEvent, CallHierarchyRequestStatus.Failed);
+            return undefined;
+        }
+
+        // Listen to a cancellation for this request. When this request is cancelled,
+        // use a local cancellation source to implicitly cancel a token.
+        let requestCanceled: CancellationSender | undefined;
+        const cancelSource: vscode.CancellationTokenSource = new vscode.CancellationTokenSource();
+        this.disposables.push(token.onCancellationRequested(() => {

```

Comment:
If a new instances of these are created every time the operation is called, a lot of disposables will accumulate and a ton of them may need to be disposed of when the provider is disposed.  Instead, could you set aside these two disposables and dispose them on subsequent iterations?  That would avoid having an indeterminate number of them lingering.

Reply:
`private dispose()`  is called after each request. But this maybe dispose newly created listensers. I'll change it to locally track the listeners.

Reply:
Wouldn't calling dispose() there be a bug, as a subsequent request may already have started?  You could be disposing of the new ones, while they're still needed.

Reply:
Yes, that's why I created local disposables instead of adding new ones in a class variable.